pallets/parachain-staking/src/lib.rs

```diff
@@ -839,25 +837,63 @@ pub mod pallet {
 		/// - The transaction's complexity is mainly dependent on updating the
 		///   `SelectedCandidates` storage in `select_top_candidates` which in
 		///   return depends on the number of `MaxSelectedCandidates` (N).
-		/// - For each N, we read `CollatorState` from the storage.
+		/// - For each N, we read `CandidatePool` from the storage.
 		/// ---------
-		/// Weight: O(N) where N is `MaxSelectedCandidates` bounded by
-		/// `MaxTopCandidates`
-		/// - Reads: MaxSelectedCandidates, TopCandidates, N * CollatorState
+		/// Weight: O(N + D) where N is `MaxSelectedCandidates` bounded by
+		/// `MaxTopCandidates` and D is the number of delegators of a
+		/// candidate bounded by `MaxDelegatorsPerCollator`.
+		/// - Reads: MaxSelectedCandidates, TopCandidates, N * CandidatePool
 		/// - Writes: MaxSelectedCandidates
 		/// # </weight>
-		#[pallet::weight(<T as pallet::Config>::WeightInfo::set_max_selected_candidates(*new, (*new).saturating_mul(T::MaxDelegatorsPerCollator::get())))]
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::set_max_selected_candidates(
+			*new,
+			T::MaxDelegatorsPerCollator::get()
+		))]
 		pub fn set_max_selected_candidates(origin: OriginFor<T>, new: u32) -> DispatchResultWithPostInfo {
 			ensure_root(origin)?;
 			ensure!(new >= T::MinCollators::get(), Error::<T>::CannotSetBelowMin);
-			let old = <MaxSelectedCandidates<T>>::get();
+			ensure!(new <= T::MaxTopCandidates::get(), Error::<T>::CannotSetAboveMax);
+			let old = MaxSelectedCandidates::<T>::get();
 
 			// *** No Fail beyond this point ***
 
-			<MaxSelectedCandidates<T>>::put(new);
-
-			// update candidates for next round
-			let (num_collators, num_delegators, _, _) = Self::update_total_stake();
+			MaxSelectedCandidates::<T>::put(new);
+
+			// Update total amount at stake for new top collators and their delegators
+			let (num_collators, num_delegators) = match new {
```

Comment:
This whole match statement looks quite complicated to me. 🤔 

I think this could maybe be simpler?

```rust
let start = old.min(new);
let end = old.max(new);

let candidates = TopCandidates::<T>::get();

let diff_delegation = 0;
let diff_collation = 0;
for candidate in candidates.skip(start).take(end - start) {
  if let Some(state) = CandidatePool::<T>::get(&candidate.owner) {
    diff_collation.saturating_add(state.stake);
    diff_delegation.saturating_add(state.total - state.stake);
  }
}

TotalCollatorStake::<T>::mutate(...)
```

Reply:
Thanks. I am very unhappy with the readability of the entire block. Your proposal looks much cleaner.